File: 0236-lowest-common-ancestor-of-a-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        parentMap = dict()
        parentMap[root] = None
        def preOrderTraversal(node):
            if node == None:
                return
            if node.left:
                parentMap[node.left] = node
            if node.right:
                parentMap[node.right] = node
            preOrderTraversal(node.left)
            preOrderTraversal(node.right)
        
        preOrderTraversal(root)
        
        pPath = []
        cur = p
        while cur:
            pPath.append(cur)
            cur = parentMap[cur]
        qPath = []
        cur = q
        while cur:
            qPath.append(cur)
            cur = parentMap[cur]
        
        i, j = len(pPath) - 1, len(qPath) - 1
        
        while pPath[i].val == qPath[j].val:
            i -= 1
            j -= 1
        
        return pPath[i+1]

File: 0236-lowest-common-ancestor-of-a-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    bool search(TreeNode* root, TreeNode* target, vector<TreeNode*> & ancestors)
    {
        if(root == nullptr) return false;
        // cout<<"root "<<root->val<<endl;
        if(root == target)
        {
            ancestors.push_back(root);
            return true;
        }
        
        bool found = false;
        if(root->left != nullptr)
        {
            found = search(root->left, target, ancestors);
        }
        if(found == false && root->right != nullptr)
        {
            found = search(root->right, target, ancestors);
        }
        
        if(found){ancestors.push_back(root);}
        return found;
    }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) 
    {
        vector<TreeNode*> pAncestors;
        vector<TreeNode*> qAncestors;
        
        // cout<<"p\n";
        search(root, p, pAncestors);
        
        // cout<<"q\n";
        search(root, q, qAncestors);
        
//         for(auto ancestor: pAncestors) cout<< ancestor->val<<" ";
//         cout<<endl;
//         for(auto ancestor: qAncestors) cout<<ancestor->val<<" ";
//         cout<<endl;
        
        int index1, index2;
        index1 = pAncestors.size()-1;
        index2 = qAncestors.size()-1;
        
        while(index1 >= 0 && index2 >= 0 && pAncestors[index1] == qAncestors[index2])
        {
            index1--; index2--;
        }
        
        

        
        return pAncestors[index1 + 1];
    }
};
